### PR TITLE
Adding energy strain in the solids of the StructuralMechaniicsApplication

### PR DESCRIPTION
**Pull request by loumalouomega**
Suggested by @annasophie91

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h
Comment (rubenzorrilla) [inline applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h]: Is it strictly necessary to store the strain energy as a member variable? I'm just asking since this is not a good practice from the memory consumption point of view... As far as I can understand, you do it to above the stress computation twice, isn't it?

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_strain.h
Comment (rubenzorrilla) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_strain.h]: See above

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_stress.h
Comment (rubenzorrilla) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_stress.h]: See above again

**Comment by RiccardoRossi** [comment]
Guys, this should not be done this way. See the recent discussion on CLs

**Comment by loumalouomega** [comment]
I just mimic the behaviour in the SolidMechanics, I will solve it

**Comment by loumalouomega** [comment]
@RiccardoRossi How do I access to the parameters in the GetValue method?, I can not calculate anything without them

**Comment by RiccardoRossi** [comment]
the idea was to add a new Calculate method which allowed computing the strain energy when needed.
this allowed not to store the extra double per every gauss point (which is definitely to be avoided)

**Comment by loumalouomega** [comment]
Yes, but this methods does exist right now?

**Comment by loumalouomega** [comment]
To do what you say you need to map the GetValue, but the input of the GetValue and CalculateValue are not the same:

~~~cpp
    /**
     * calculates the value of a specified variable
     * @param rParameterValues the needed parameters for the CL calculation
     * @param rThisVariable the variable to be returned
     * @param rValue a reference to the returned value
     * @param rValue output: the value of the specified variable
     */    
    virtual int& CalculateValue(Parameters& rParameterValues, const Variable<int>& rThisVariable, int& rValue);
    /**
     * returns the value of a specified variable
     * @param rThisVariable the variable to be returned
     * @param rValue a reference to the returned value
     * @param rValue output: the value of the specified variable
     */
    virtual int& GetValue(const Variable<int>& rThisVariable, int& rValue);
~~~

**Comment by loumalouomega** [comment]
The thing I can do is to use CalculateValue instead of GetValue

**Comment by RiccardoRossi** [comment]
correct, that's the way

**Comment by loumalouomega** [comment]
Ok, done

File: applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp]: what was wrong with the name rValues?

File: applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp]: to be honest i am not sure of this 0.5. Are u sure about that?

File: applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp]: to add on this, by changing that name you multiplyied by a factor 100 the effort needed to revise the PR ... so please don't...

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_strain.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_strain.cpp]: same ... please do not change the rValues "just because"

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_stress.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_stress.cpp]: you don't need to check this option. you already have the other if to control that ( i think that COMPUTE_STRAIN_ENERGY should fly out)

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: you don't need to set this flag.

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_strain.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_strain.cpp]: imagen

This is from constitutive_law.h

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_strain.cpp]: Ok, I just saw that the name in constitutive_law. h for calculate mnaterial reponse was right:

imagen

I will revert it, but maybe it is a little bit inconsistent

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_stress.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_stress.cpp]: OK

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: OK

File: applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp]: Sorry I check it in the base class, but I checked wrong

File: applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp]: imagen
From Belytschko

**Comment by loumalouomega** [comment]
@RiccardoRossi Done, reverting the names and applying the changes suggested, sorry I look wrongly in the base class

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_strain.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_strain.cpp]: same here... on all of those rParamterValues

**Comment by RiccardoRossi** [comment]
i think the rParamtersValue is still there in one of the classes.

i'll approve this assuming that you'll fix it. However please wait for the checks to be passed before merging

**Review by RiccardoRossi — APPROVED** [review]
approving provided that you fix the remaining rParamterValues

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_strain.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_strain.cpp]: Ok

**Pull request merged**
